src/graphql_clj/visitor.clj

```diff
@@ -112,6 +129,36 @@
         (assoc :state (-> result :document section :state))
         (update-in [:document section] dissoc :state))))
 
+(defn- schema-definition
+  "Given a parsed schema document, return the schema definition that defines the entry point
+   types (schema and optionally mutation). Can return nil, in which case the default type
+   `Query` will be used and mutations will be disabled."
+  [parsed-schema]
+  (some->> parsed-schema
+           :type-system-definitions
+           (filter #(= :schema-definition (:node-type %)))
+           first))
+
+(defn- upsert-schema-entrypoint
+  [type-system-definitions root-query-name]
+  (let [grouped (group-by #(= root-query-name (:type-name %)) type-system-definitions)
+        entrypoint (intro/upsert-root-query (first (get grouped true)) root-query-name)
+        definitions-without-entrypoint (get grouped false)]
+    (conj definitions-without-entrypoint entrypoint)))
+
+(defn- root-fields
+  "Given a parsed schema document, return [root-name {root-field Type}]
+   When validating queries, we need to know the types of the root fields to map to the same specs
+   that we registered when parsing the schema.  Similar for mutations."
+  [root-name parsed-schema]
+  (some->> parsed-schema
+           :type-system-definitions
+           (filter #(= (:type-name %) root-name))
+           first
+           :fields
+           (map (juxt (comp box/box->val :field-name) (comp box/box->val :type-name)))
```

Comment:
Fixed this for list types.  I think non-null types are working (there is a test for this with loremIpsum of type `String!`) but I may be missing something.